Port Pileup

magicDGS opened the pull request:
Porting Pileup (https://github.com/broadinstitute/gatk/issues/1581) and tests

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/Pileup.java:
```diff
@@ -0,0 +1,176 @@
+ *
+ * <h3>Usage example</h3>
+ * <pre>
+ * java -jar GenomeAnalysisTK.jar \
```
can you update the doc to the new use?

magicDGS replied:
Done

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/Pileup.java:
```diff
@@ -0,0 +1,176 @@
+        + "additional information can be added to the output as extra columns; see options detailed below.",
+    oneLineSummary = "Print read alignments in Pileup-style format",
+    programGroup = QCProgramGroup.class)
+public class Pileup extends LocusWalker {
```
make it final class

magicDGS replied:
Done

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/Pileup.java:
```diff
@@ -0,0 +1,176 @@
+    programGroup = QCProgramGroup.class)
+public class Pileup extends LocusWalker {
+
+    private static final String verboseDelimiter = "@"; // it's ugly to use "@" but it's literally the only usable character not allowed in read names
```
constants use THIS_STYLE for names

magicDGS replied:
Done

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/Pileup.java:
```diff
@@ -0,0 +1,176 @@
+     * and read mapping quality.  These per read items are delimited with an '@' character.
+     */
+    @Argument(fullName = "showVerbose", shortName = "verbose", doc = "Add an extra verbose section to the pileup output", optional = true)
+    public boolean SHOW_VERBOSE = false;
```
fields use `thisStyle` for names

magicDGS replied:
Done

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/Pileup.java:
```diff
@@ -0,0 +1,176 @@
+     * annotated with the corresponding source ROD identifier.
+     */
+    @Argument(fullName = "metadata", shortName = "metadata", doc = "ROD file containing metadata", optional = true)
+    public List<FeatureInput<Feature>> rods = new ArrayList<>();
```
ditch the `rod` terminology everywhere

magicDGS replied:
Done

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/Pileup.java:
```diff
@@ -0,0 +1,176 @@
+    @Argument(fullName = "outputInsertLength", shortName = "outputInsertLength", doc = "Output insert length", optional = true)
+    public boolean outputInsertLength = false;
+
+    private int result = 0;
```
document what this is or delete - i don't think it's useful for anything

magicDGS replied:
Deleted

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/Pileup.java:
```diff
@@ -0,0 +1,176 @@
+        final StringBuilder s = new StringBuilder();
+        s.append(
+            String.format("%s %s", basePileup.getPileupString((hasReference()) ? (char) ref.getBase() : 'N'), rods));
+        if (outputInsertLength)
```
add braces `{` and `}` to all `if` and `for` etc

magicDGS replied:
Done

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/Pileup.java:
```diff
@@ -0,0 +1,176 @@
+     * @return String representation of the reference-ordered data.
+     */
+    private String getReferenceOrderedData(final FeatureContext tracker) {
+        ArrayList<String> rodStrings = new ArrayList<>();
```
declare temp using the interface `final List<String>` - the only legitimate use for `ArrayList` is in `new ArrayList<>`

magicDGS replied:
Done

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/Pileup.java:
```diff
@@ -0,0 +1,176 @@
+            rodStrings.add(datum.toString());
+        }
+        String rodString = Utils.join(", ", rodStrings);
+        if (!rodString.equals(""))
```
braces

magicDGS replied:
Done

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/Pileup.java:
```diff
@@ -0,0 +1,176 @@
+    private static String insertLengthOutput(final ReadPileup pileup) {
+        final Integer[] insertSizes = new Integer[pileup.size()];
+        int i = 0;
+        for (final PileupElement p : pileup) {
```
turn this method into a 1-liner using streams and `Collectors.joining`

magicDGS replied:
Done

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/Pileup.java:
```diff
@@ -0,0 +1,176 @@
+        sb.append(pileup.getNumberOfDeletions());
+        sb.append(" ");
+        for (final PileupElement p : pileup) {
+            if (isFirst)
```
braces

magicDGS replied:
Done

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/Pileup.java:
```diff
@@ -0,0 +1,176 @@
+
+    @Override
+    public Object onTraversalSuccess() {
+        out.println("[REDUCE RESULT] Traversal result is: " + result);
```
remove this - it's useless

magicDGS replied:
Done

akiezun commented on src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java:
```diff
@@ -235,6 +235,16 @@ public String toString() {
                 getQualsString());
     }
 
+    public String getPileupString(final Character ref) {
```
document what this method does, change `Character` to `char` and add a unit test for it

magicDGS replied:
Done

akiezun commented on src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java:
```diff
@@ -255,6 +265,13 @@ public int getNumberOfElements(final Predicate<PileupElement> peFilter){
+	/**
+     * Shortcut for get the number of deletions
+     */
+    public int getNumberOfDeletions() {
```
why do we need this method? can you just inline it in the caller? If it's used just once let's either inline or keep it private.

magicDGS replied:
I've just add it as a shortcut, because I'm also using it instead of using the lambda expression. For API users it would be nice, but I could remove it if you think that is not going to be useful.

akiezun commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/qc/PileupIntegrationTest.java:
```diff
@@ -0,0 +1,28 @@
+				" -metadata " + TEST_DATA_DIRECTORY + "feature_data_source_test.vcf" +
+				" -outputInsertLength" +
+				" -O %s",
+			Arrays.asList(TEST_OUTPUT_DIRECTORY + "expectedPileup.txt")
```
how was the expected output created? Did you use GATK3.5? Can you add a comment?

akiezun replied:
I think you should use gatk3.5 and change the output to remove `ROD` and `[REDUCE RESULT]`

akiezun replied:
put the expected file in a `Pileup` subdirectory

magicDGS replied:
I run `Pileup`with GATK3.5 to generate the expectedPileup.txt, and that's why the test fail for me: `VariantContext` that overlaps have "metadata" as source in GATK3.5, but with FeatureContext.getValues() is "Unknown". Should I modify also that in the GATK3.5 output?

magicDGS replied:
I think I did it correctly.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/Pileup.java:
```diff
@@ -0,0 +1,176 @@
+            insertSizes[i] = p.getRead().getFragmentLength();
+            ++i;
+        }
+        return Utils.join(",", insertSizes);
```
this does not even compile for me (we treat warnings as errors): 
```
/Users/akiezun/IdeaProjects/gatk/src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/Pileup.java:142: warning: non-varargs call of varargs method with inexact argument type for last parameter;
        return Utils.join(",", insertSizes);
```

magicDGS replied:
Using `Collectors.joining` may solve the issue.

akiezun commented:
back to @magicDGS - good start, some comments to work on

magicDGS commented:
Back to you, @akiezun. I think that I did almost everything that you mentioned, and I refactored names in the tool class to fit in the new framework.

I believe that for some API users it would be nice to have shortcuts like the `getNumberOfDeletions()`, so I didn't remove it. But I can do it in a later commit if you really think that is unnecessary.

akiezun commented on src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java:
```diff
@@ -256,6 +272,13 @@ public int getNumberOfElements(final Predicate<PileupElement> peFilter){
     /**
+     * Shortcut for get the number of deletions
+     */
+    public int getNumberOfDeletions() {
```
I'm not a fan of such methods (trivial for clients to code themselves and begs to have a `getNumberOfElementsBeforeInsertion`, `getNumberOfElementsAfterDeletionEnd` etc - why not have one for every boolean method on `PileupElement`?). If you really want to keep it, add unit tests for it.

magicDGS replied:
Removed.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/Pileup.java:
```diff
@@ -0,0 +1,178 @@
+            featuresStrings.add(datum.toString());
+        }
+        String featuresString = Utils.join(", ", featuresStrings);
+        if (!featuresString.equals("")) {
```
replace `featuresString.equals(""))` with `featuresString.isEmpty())`

magicDGS replied:
Done (plus rewritting with stream).

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/Pileup.java:
```diff
@@ -0,0 +1,178 @@
+     * @return String representation of the metadata
+     */
+    private String getFeaturesString(final FeatureContext featureContext) {
+        List<String> featuresStrings = new ArrayList<>();
```
make this variable final

magicDGS replied:
Don't needed anymore because now is a lambda expression.

akiezun commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/qc/PileupIntegrationTest.java:
```diff
@@ -0,0 +1,28 @@
+import java.io.IOException;
+import java.util.Arrays;
+
+public class PileupIntegrationTest extends CommandLineProgramTest {
```
this test is good but not extensive enough. method `createVerboseOutput` is not tested at all.
Test also the case without a reference

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/Pileup.java:
```diff
@@ -0,0 +1,178 @@
+     * @param pileup the pileup to format
+     * @return a comma-separated string with insert lengths
+     */
+    private static String insertLengthOutput(final ReadPileup pileup) {
```
change into a package-visible method, add `@VisibleForTesting` and add unit tests for it

magicDGS replied:
Done

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/Pileup.java:
```diff
@@ -0,0 +1,178 @@
+     * @param pileup the pileup to format
+     * @return formatted string with read information for the pileup
+     */
+    private static String createVerboseOutput(final ReadPileup pileup) {
```
change into a package-visible method, add `@VisibleForTesting` and add unit tests for it

magicDGS replied:
Done

akiezun commented on src/test/resources/org/broadinstitute/hellbender/tools/walkers/qc/pileup/expectedPileup.txt:
```diff
@@ -0,0 +1,238 @@
+1 200 N A B [Feature(s): [VC Unknown @ 1:199-200 Q232.46 of type=INDEL alleles=[GG*, G] attr={} GT=[], [VC Unknown @ 1:200 Q232.46 of type=SNP alleles=[G*, A] attr={} GT=[]] -130
```
hmm, every single reference base here seems to be `N` - can you pick inputs so that actual reference bases (non-N ones) would show up here?

magicDGS replied:
I'm struggling a lot with the test files, because there are a lot of them and I don't know how they are organized. I've already have problems implementing some tests without finding the correct reference. Do you have any suggestion, @akiezun? Which set of files could I use for the tests?

akiezun commented:
back to @magicDGS - second review done. getting there!

magicDGS commented:
Included everything except changes in `PileupIntegrationTest`. @akiezun, could you let me know what kind of files could I use for implement the integrations test?

akiezun commented:
sure, you should use files in `src/test/resources/large`. Here's an example using GATK3.5:
```
java -jar ~/bin/GenomeAnalysisTK-3.5.0/GenomeAnalysisTK.jar -T Pileup -R src/test/resources/large/human_g1k_v37.20.21.fasta -I ./src/test/resources/large/CEUTrio.HiSeq.WGS.b37.NA12878.20.21.bam -L 20:9999990-9999999 -L 20:9999900-9999910
```

output:
```
20 9999902 T T D
20 9999903 G GGG EE:
20 9999904 T TTT DIE
20 9999905 T TTTTTT EGCAGE
20 9999906 A AAAAAA FGCDG=
20 9999907 A AAAAAAA GF>EGDF
20 9999908 A AAAAAAA GG>EGEG
20 9999909 A AAAAAAAAA HE4CGEFAA
20 9999910 T TTTTTTTTTTT FFDCHBDDDDD
20 9999990 T TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT GFGGFFFFGGFEGGGGAGGGGFFGFG<GGGGG@GGFFBFFGGF:FFFC5?ECFGFFGFFFDF3:GFGG#FCFFFC9EGADB@EA
20 9999991 A AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAATAAAAAAAAAAACAAAAAAAA GBFGFCFGGGGDGFFGFFFGGGFGFC9F9FFDFFFGGFGGEFG<GGGD9FHDGGGFFGGGEG5-FGFF#FFFGFF+EGE?DBFD
20 9999992 C CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC EFFEGFFEEEEBEGGECGGEEEGEGG9G8GGGBGGEEEEEGGE@EEEACGFEEGEEGEEEBE',GEGG#EEEEEE,DHDCCBDBD
20 9999993 C CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC GFFGFFFFFGFGGFFGDGFGGFEBDG?G<GGGAGGGFCFFGGFDFFFFGGGEFGFFGFBFFF3;GFGG#FGFFFG;FGFEEEFCD
20 9999994 A AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA HFFHFDCGHHHEHFEHCFFHHHC9FF3F<FFE@FFHHDGHFFH7AGGCHFHEHGGGDHGCAH7:FGFF#GGGHGGDGFGFFFGBE
20 9999995 C CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC FFFEFEEEEEEEEFFEBGGEEEDAFG9B=GGGDFGEEDEEGGE<EEEEGGFEEGEEGEEEFE9/GEGG#EEEEEFCDGCDDCDACF
20 9999996 A AAAAAAAAAAAAAAAAAAAAAAAAAACAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA HFFHFDEGHHHEHEDHEFFHHHFEEF&FFFFEEFFHGEGHFGH;GGGCH?HFHFGGFHGGGH:4FGFG#GGGHGGBEGFGFFGEG?
20 9999997 G GGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGTGGGGGGGGGGGGGGGGGGGGG GDDEDB>GGGGEG?AGACBGGGBCBA2B=BCCFCBGGGGG=CG=DGGGG3GDGCGG?GGGEG;+CGCC#GGGFGH:GCGFFFFEFCABEB
20 9999998 T TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT FGBFGD?FEGFHF?DGFEEEGFCGEBAEBEEEBEEFGBFGAEF)FFFFA9DEFEGFBFFFGF3<EFFF#FDFFFD=FBDFFFFEFBDCDB
20 9999999 T TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT GFFFGC>DGGHGGACGAEEGGHDGED<D@DEDGEFGGEGGBEH3GGGGG=GFGFGGBGGGGG:7EGFF#GGFGFG8FEFFEFFFFDDDCB
```

akiezun commented:
back to you @magicDGS

magicDGS commented:
I added integration tests for simple output and including features or verbose. While doing it, I realized that GATK 3.5 included some filters that wasn't included here, and that indels weren't tracked, so I changed also the code to fit the previous implementation.

Back to you @akiezun.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/Pileup.java:
```diff
@@ -0,0 +1,195 @@
+                basePileup.getPileupString((hasReference()) ? (char) referenceContext.getBase() : 'N'),
+                features));
+        if (outputInsertLength) {
+            s.append(" ").append(insertLengthOutput(basePileup));
```
add a test case that reaches this line ` s.append(" ").append(insertLengthOutput(basePileup));`

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/Pileup.java:
```diff
@@ -0,0 +1,195 @@
+     */
+    @VisibleForTesting
+    static String insertLengthOutput(final ReadPileup pileup) {
+        return pileup.getReads().stream()
```
add an integration test that reaches this method

akiezun commented:
@magicDGS this looks good. 2 more tests to add, then rebase and squash and it'll be ready to merge

akiezun commented:
@magicDGS the tests fail - maybe you misplaced the expected files (?)

magicDGS commented:
In my computer it is passing, and I think that it is because the package name for the tests is with lower case and the one returned by `BaseTest.getToolTestDataDir()` is with upper case. Because my system is case-insensitive, I can't commit with git a change in the package name (that it will be better for consistency with the rest of the tests), and that's why it also pass. 

I've just made a commit with a `toLowerCase()` call to check if it solves the problem. If it pass, let me know if I can rebase and squash, @akiezun.

akiezun commented:
ok looks fine. can you rebase and squash the commits?

magicDGS commented:
Squashed, rebased and added integration test for `--outputInsertLength`. Back to you @akiezun for (final?) review.

akiezun commented:
👍 thanks @magicDGS. good job!
